src/pam_google_authenticator.c

```diff
@@ -1493,6 +1513,125 @@ static int check_timebased_code(pam_handle_t *pamh, const char*secret_filename,
   return 1;
 }
 
+/* 
+ * Add a 'config' variable that says we logged in from a particular place
+ * at a particular time.  Only remembers the last 10 logins, which 
+ * are replaced in LRU order.
+ *
+ * Returns 0 on success.
+ */
+int
+update_logindetails(pam_handle_t *pamh, const Params *params, char **buf) {
+  const char *rhost = get_rhost(pamh, params);
+  const time_t now = get_time();
+  time_t oldest = now;
+  int oldest_index = 0;
+  int i;
+  char *line;
+  unsigned long when;
+  char name[] = "LAST ";
+  char *host;
+
+  if (rhost == NULL) {
+    return -1;
+  }
+
+  for (i = 0; i < 10; i++) {
+    name[4] = i + '0';
+    line = get_cfg_value(pamh, name, *buf);
+    if (line == &oom) {
+      /* Fatal! */
+      return -1;
+    }
+
+    if (!line) {
+      /* Make first empty line the oldest */
+      if (oldest) {
+        oldest_index = i;
+        oldest = 0;
+      }
+      continue;
+    }
+
+    if (sscanf(line, " %m[0-9a-fA-F:.] %lu ", &host, &when) != 2) {
+      log_message(LOG_ERR, pamh, "Malformed LAST%d line\n", i);
+      continue;
+    }
+
+    if (!strcmp(host, rhost)) {
+      break;
+    }
+        
+    if (when < oldest) {
+      oldest_index = i;
+      oldest = when;
+    }
+
+    free(line);
+    free(host);
+  }
+
+  if (i == 10) {
+    name[4] = oldest_index + '0';
+  } else {
+    free(line);
+    free(host);
+  }
+
+  int len = strlen(rhost) + 100;
+  char *value = calloc(len, 1);
+
+  snprintf(value, len - 1, "%s %lu", rhost, (unsigned long)now);
```

Comment:
the `-1` here is not needed. Or shouldn't be. Is this for portability?

Reply:
I know the spec says it shouldn't but some library implementations I've encountered fail to nul-terminate the string if the length is exhausted.  
Also, 23 should be enough... not 100.  ⌊64 log 2 + 1⌋ is 21, plus one for the terminating NUL and one for the space.  And we could avoid the calloc and strlen too ...

Reply:
I don't mind either the -1 nor the extra space. Better safe than sorry and yeah `snprintf` has had a… strange history.